Migrate Categories Add/Edit actions

sarjon opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR migrates Categories Add & Edit actions.
| Type?         | new feature
| Category?     | BO
| BC breaks?    | yes (overrides)
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10194
| How to test?  | Access "Categories" page in BO. Actions "Add new root category", "Add new category" and "Edit" are now migrated and should work the same as before. **NOTE: in order to keep this PR manageable, some functionality where skipped and will be implemented in another PR.**

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/10408)
<!-- Reviewable:end -->

Commit:
add template rendering

Commit:
create initial category type and render it

Commit:
add translatable textarea & group access inputs

Commit:
update form with required fields

Commit:
update webpack

Commit:
add categories js

Commit:
add initial command for add new category

Commit:
add material choice tree

Commit:
add ui interaction for choice tree

Commit:
make choice tree toggable

Commit:
add reusable category choice tree type

Commit:
update category form type with category choice tree

Commit:
update theme scss with new choice tree component

Commit:
update category js

Commit:
fix category choice tree form value

Commit:
removed non used option

Commit:
add category creation command handler

Commit:
update command data validation

Commit:
handle customer group association

Commit:
minor form update

Commit:
add support for shop association in add category command handler

Commit:
add initial shop choice tree type

Commit:
ignore null values for choice tree

Commit:
implemented initial image uploading adapter

Commit:
fix image uploader validation & add thumbnail uploader

Commit:
add input with countable length

Commit:
add translatable form type that accepts different types

Commit:
update category type to use translatable type

Commit:
extract category form into separate template

Commit:
add category editing form

Commit:
display category edit form with prefilled data

Commit:
preselect choices values in choice tree from type

Commit:
fix display of choice tree items

Commit:
wip thumnail menu image uploader

Commit:
update namespace for Category domain

Commit:
fix namepsace in EditableCategory

Commit:
finished implementing menu thumbnail uploader

Commit:
add abstract category type

Commit:
add root category form page

Commit:
add abstract command for adding category

Commit:
implement AddRootCategoryHandler

Commit:
wip AddRootCategoryCommand handling in controller

sarjon committed:
remove form type duplicate

sarjon committed:
do not add shop association input when multistore is not used

sarjon committed:
do not allow selected category as it's parent

sarjon committed:
fix max length for meta title

sarjon committed:
update add / edit layout titles

sarjon committed:
add category status description in category form

sarjon committed:
register add root category command

sarjon committed:
add thumbnail & cover images data to category form

sarjon committed:
implement image source parser

sarjon committed:
add cover, thumbnail and menu thumbnails display when editing category

sarjon committed:
handle images display when they do not exist

sarjon committed:
update add / add-root commands

sarjon committed:
copy category name to link rewrite

sarjon committed:
add query for default group names

sarjon committed:
improve choice tree js with auto children check

sarjon committed:
refactor shop tree choice provider into adapter service

sarjon committed:
keep BO category hook

sarjon committed:
implement default groups query

sarjon committed:
refactor add commands

sarjon committed:
implement category edit handlers

sarjon commented on admin-dev/themes/new-theme/js/components/form/choice-tree.js:
```diff
@@ -0,0 +1,142 @@
+
+      $itemWithChildren
+        .find('ul input[type="checkbox"]')
+        .prop('checked', $clickedCheckbox.is(':checked'));
```
@CaptainYouz when material checkbox is checked using javascript like `.prop('checked', true)` it loses animation, do you have any ideas how to workaround it? :thinking:

sarjon committed:
fix name to link rewrite copier for root category

sarjon committed:
improve category add / edit handling

sarjon committed:
update command building

sarjon committed:
fix choice tree for root category form

sarjon committed:
add exception handling

sarjon committed:
cs fixer

matks commented on admin-dev/themes/new-theme/js/pages/categories/index.js:
```diff
@@ -0,0 +1,45 @@
+  new TextWithLengthCounter();
+  new NameToLinkRewriteCopier();
+
+  new ChoiceTree('#category_id_parent');
```
👍

matks commented on admin-dev/themes/new-theme/js/pages/categories/name-to-link-rewrite-copier.js:
```diff
@@ -0,0 +1,48 @@
+export default class NameToLinkRewriteCopier {
+  constructor() {
+    ['category', 'root_category'].forEach((categoryType) => {
+      const $categoryForm = $('form[name="' + categoryType + '"]');
```
I guess @PierreRambaud would tell you to use [ES6 notation](https://developers.google.com/web/updates/2015/01/ES6-Template-Strings) here and in the whole file

PierreRambaud replied:
Yep, our eslint configuration will cry otherwise, and it's pretty much easier to read.

matks commented on src/Adapter/Category/CommandHandler/AbstractCategoryHandler.php:
```diff
@@ -0,0 +1,142 @@
+        // This is a workaround to make Category's object model work.
+        // Inside Category::add() & Category::update() method it checks if shop association is submitted
+        // by retrieving data directly from $_POST["checkBoxShopAsso_category"].
+        $_POST['checkBoxShopAsso_category'] = $command->getAssociatedShopIds();
```
😨 
I know you have to deal with the legacy code but ... this looks so bad ...

sarjon replied:
:(

matks commented on src/Adapter/Category/CommandHandler/AddCategoryHandler.php:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(AddCategoryCommand $command)
```
👍

matks commented on src/Adapter/Category/CommandHandler/AddCategoryHandler.php:
```diff
@@ -0,0 +1,57 @@
+
+        $this->uploadImages($category, $command);
+
+        return new CategoryId($category->id);
```
I'm not sure whether it's useful but ... should we check whether `new CategoryId($category->id)` is not null ? Since we use ObjectModels which are not very reliable

matks commented on src/Adapter/Category/Image/CategoryThumbnailRetriever.php:
```diff
@@ -0,0 +1,104 @@
+        $imageTag = ImageManager::thumbnail(
+            $imagePath,
+            $thumbnailCachedImageName,
+            45,
```
What is this 45 ?

matks commented on src/Adapter/Category/Image/CategoryThumbnailRetriever.php:
```diff
@@ -0,0 +1,104 @@
+        $replacement = 'src="' . $this->legacyContext->getRootUrl();
+        $imageTag = preg_replace('/src="(\\.\\.\\/)+/', $replacement, $imageTag);
+
+        preg_match( '@src="([^"]+)"@' , $imageTag, $path );
```
😝 my brain melt
Can you put a private function here with a meaningful name such as `findLegacyPath()` to hide this crazy rule ?

Do you think this rule will be reused in other parts of the code ? If yes, it needs to be put inside in an Util class

matks commented on src/Adapter/Category/QueryHandler/GetCategoryForEditingHandler.php:
```diff
@@ -0,0 +1,197 @@
+
+        $thumb = '';
+        $imageTag = '';
+        $formatted_small = ImageType::getFormattedName('small');
```
camelCase variable 😉

matks commented on src/Adapter/Category/QueryHandler/GetCategoryForEditingHandler.php:
```diff
@@ -0,0 +1,197 @@
+        if (!is_file($thumb)) {
+            $thumb = $image;
+            $imageTag = ImageManager::thumbnail($image, 'category_' . $categoryId->getValue() . '-thumb.jpg', 125, 'jpg', true, true);
+            ImageManager::resize(_PS_TMP_IMG_DIR_ . 'category_' . $categoryId->getValue() . '-thumb.jpg', _PS_TMP_IMG_DIR_  . 'category_' . $categoryId->getValue() . '-thumb.jpg', (int) $image_type['width'], (int) $image_type['height']);
```
Line too long

matks commented on src/Adapter/Category/QueryHandler/GetCategoryForEditingHandler.php:
```diff
@@ -0,0 +1,197 @@
+            if ($formatted_small == $image_type['name']) {
+                $thumb = _PS_CAT_IMG_DIR_ . $categoryId->getValue() . '-' . $image_type['name'] . '.jpg';
+                if (is_file($thumb)) {
+                    $imageTag = ImageManager::thumbnail(
```
Code does not look pretty but I guess this comes from the legacy, right ?

sarjon replied:
yup, the legacy code is very hard to understand, so i try to encapsulate it in command as much as i can without modifying it and introducing new issues. :/

matks commented on src/Adapter/Group/QueryHandler/GetDefaultGroupsHandler.php:
```diff
@@ -0,0 +1,94 @@
+    private $contextLangId;
+
+    /**
+     * @param ConfigurationInterface $configuration
```
missing 1 parameter in phpDoc

matks commented on src/Adapter/Image/Uploader/AbstractImageUploader.php:
```diff
@@ -0,0 +1,77 @@
+use Tools;
+
+/**
+ * Class AbstractImageUploader encapsulates reusable legacy methods used for image uploading.
```
👍

matks commented on src/Adapter/Image/Uploader/CategoryCoverImageUploader.php:
```diff
@@ -0,0 +1,136 @@
+use ImageManager;
+use ImageType;
+use PrestaShop\PrestaShop\Core\Image\Uploader\Exception\ImageOptimizationException;
+use PrestaShop\PrestaShop\Core\Image\Uploader\Exception\ImageUploadException;
```
👍

matks commented on src/Adapter/Image/Uploader/CategoryMenuThumbnailUploader.php:
```diff
@@ -0,0 +1,109 @@
+        $availableKeys = array_diff($allowedKeys, $usedKeys);
+
+        // HelperImageUploader::process expects
+        $_FILES['thumbnail'] = [
```
😭 
(I know it's not your fault)

sarjon replied:
it's really sad. :(

PierreRambaud replied:
:crying_cat_face:

matks commented on src/Adapter/Image/Uploader/CategoryMenuThumbnailUploader.php:
```diff
@@ -0,0 +1,109 @@
+
+        // HelperImageUploader::process expects
+        $_FILES['thumbnail'] = [
+            'error' => [$uploadedImage->getError()],
```
😭 😭 😭

PierreRambaud replied:
:crying_cat_face: :crying_cat_face: :crying_cat_face:

matks commented on src/Adapter/Image/Uploader/CategoryMenuThumbnailUploader.php:
```diff
@@ -0,0 +1,109 @@
+        $uploadedFiles = $helper->process();
+
+        if (count($availableKeys) < count($files)) {
+            //@todo: throw exception
```
let's do it !

matks commented on src/Adapter/Image/Uploader/CategoryMenuThumbnailUploader.php:
```diff
@@ -0,0 +1,109 @@
+
+            // Evaluate the memory required to resize the image: if it's too much, you can't resize it.
+            if (isset($uploadedFile['save_path']) && !ImageManager::checkImageMemoryLimit($uploadedFile['save_path'])) {
+                //@todo: throw exception
```
yes we can

matks commented on src/Adapter/Image/Uploader/CategoryMenuThumbnailUploader.php:
```diff
@@ -0,0 +1,109 @@
+                unlink($uploadedFile['save_path']);
+            }
+
+            //Necesary to prevent hacking
```
`Necessary`

matks commented on src/Core/Domain/Category/Command/AbstractRootCategoryCommand.php:
```diff
@@ -0,0 +1,363 @@
+    /**
+     * @var string[]
+     */
+    private $name;
```
Why is it `$name` instead of `$names` if this is an array ?

sarjon replied:
well, i had idea of naming it as `$names` but is it really a correct way of doing this, isnt it confusing?

does this mean that Category can have many different names? or does it mean that category has single name, but in different languages? :thinking: see my point? either way, im not against `$names` if that makes more sense for you. :)

matks commented:
In another PR, I will ask you to write unit tests for command objects. These unit tests should check:
- the objects can be constructed
- check that "classic" invalid argument usecases throw exception
- check interfaces are implemented

These tests are not intended to find bugs now but rather to allow modifications later: if we break basic behavior they should fail (for ex. if we update an Interface without updating the object implementing it)

matks commented on src/Core/Domain/Category/Exception/CategoryConstraintException.php:
```diff
@@ -0,0 +1,63 @@
+class CategoryConstraintException extends CategoryException
+{
+    /**
+     * Code is used when category does not have name.
```
👍

matks commented on src/Core/Domain/Category/ValueObject/CategoryId.php:
```diff
@@ -0,0 +1,72 @@
+    /**
+     * @param int $categoryId
+     */
+    private function setCategoryId($categoryId)
```
What is the need for a private setter ?

PierreRambaud replied:
If you want to change the id, you need to generate a new class.

matks commented on src/Core/Grid/Definition/Factory/AbstractGridDefinitionFactory.php:
```diff
@@ -62,6 +62,10 @@
             'definition' => $definition,
         ]);
 
+        \Hook::exec('action' . $definition->getId() . 'GridDefinitionModifier', [
```
Why do we use both `$this->hookDispatcher->dispatchWithParameter()` and `\Hook::exec()` ? Should not it be either one or the other ?

sarjon replied:
using `\Hook::exec()` works as expected. :D but i unintentionally left it here when playing with hooks, gonna remove it. :)

matks commented on src/Core/Image/Parser/ImageTagSourceParser.php:
```diff
@@ -0,0 +1,70 @@
+ * This service helps retrieving image path from image tag generated by legacy ImageManager::thumbnail() method
+ * so image can be displayed in new pages.
+ */
+final class ImageTagSourceParser implements ImageTagSourceParserInterface
```
This is a great candidate for unit tests. Easy to isolate, easy dependency mocking. Input and outputs are simple.

matks commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -0,0 +1,420 @@
+            return $this->handleConstraintException($exception);
+        }
+
+        $messages = [
```
Can we rename it `availableMessagesForDisplay` or something like that ? It took me a few minutes to understand the `if (isset($messages[$type]))` logic so maybe we can make it a bit more understandable

PierreRambaud commented on src/Core/Image/Parser/ImageTagSourceParser.php:
```diff
@@ -0,0 +1,70 @@
+    public function parse($imageTag)
+    {
+        $replacement = 'src="/';
+        $imageTag = preg_replace('/src="(\\.\\.\\/)+/', $replacement, $imageTag);
```
Why `\\` only one is needed otherwise you escape the `\` and not the `.`

PierreRambaud replied:
and same as previous broken regex.

PierreRambaud commented on admin-dev/themes/new-theme/js/components/form/choice-tree.js:
```diff
@@ -0,0 +1,142 @@
+      return;
+    }
+
+    if ($parentWrapper.hasClass('collapsed')) {
```
This function says if you're not in `expanded` context you're in `collapsed` context, this condition look useless.

sarjon replied:
not really, because if `li` element does not have any children then its neither in `expanded` not `collapsed` context.

PierreRambaud replied:
You need to add a test to be sure there is a closest li to be explicit please.

PierreRambaud commented on admin-dev/themes/new-theme/js/components/form/text-with-length-counter.js:
```diff
@@ -0,0 +1,53 @@
+ */
+export default class TextWithLengthCounter {
+  constructor() {
+    $(document).find('.js-text-with-counter-input-group').on('input', 'input[type="text"]', (e) => {
```
`$(document).on('input', '.js-text-with-counter-input-group input[type="text"]'` will be faster

PierreRambaud commented on admin-dev/themes/new-theme/js/pages/categories/name-to-link-rewrite-copier.js:
```diff
@@ -0,0 +1,48 @@
+
+        $categoryForm
+          .find('input[name="' + categoryType + '[link_rewrite][' + langId + ']"]')
+          .val(str2url($nameInput.val(), 'UTF-8'));
```
What is `str2url`? Idon't see any import

sarjon replied:
it comes from `ps-root/js/admin.js` which is included in every BO page, so its a global function. i know its not a clean way, but is there a better way to deal with it? :thinking: 

i saw this function was used the same way in other new JS file. :/

PierreRambaud replied:
Could you have the time to add it? in `admin-dev/themes/new-theme/js/components` ?

PierreRambaud commented on src/Adapter/Category/CommandHandler/AddRootCategoryHandler.php:
```diff
@@ -0,0 +1,85 @@
+    public function handle(AddRootCategoryCommand $command)
+    {
+        $category = new Category();
+        $category->is_root_category = 1;
```
`is_` means it should be a boolean no?

sarjon replied:
in theory yes, but this object model fields are directly linked to db table, so it's 1 either way. :/

PierreRambaud replied:
If you speak about:

```sql
`is_root_category` tinyint(1) NOT NULL DEFAULT '0',`
```
This is the way MySQL declare boolean.. Yeah, f**k logic, so you can set value to `true`.


More, the ObjectModel says: 
```php
    /** @var bool is Category Root */
    public $is_root_category;
```

PierreRambaud commented on src/Adapter/Category/Image/CategoryThumbnailRetriever.php:
```diff
@@ -0,0 +1,104 @@
+     */
+    public function retrieve($categoryId)
+    {
+        $imageType = 'jpg';
```
const or options please.

PierreRambaud commented on src/Adapter/Category/Image/CategoryThumbnailRetriever.php:
```diff
@@ -0,0 +1,104 @@
+        $replacement = 'src="' . $this->legacyContext->getRootUrl();
+        $imageTag = preg_replace('/src="(\\.\\.\\/)+/', $replacement, $imageTag);
+
+        preg_match('@src="([^"]+)"@', $imageTag, $path);
```
see this regex twice at least, should be a const or a generic function.

PierreRambaud commented on src/Adapter/Category/Image/CategoryThumbnailRetriever.php:
```diff
@@ -0,0 +1,104 @@
+
+        // because legacy uses relative path to reach a directory under root directory...
+        $replacement = 'src="' . $this->legacyContext->getRootUrl();
+        $imageTag = preg_replace('/src="(\\.\\.\\/)+/', $replacement, $imageTag);
```
Broken regex. you don't escape the `/`, and `\\` means you match the character `\` literally.

Should be : `src="(\.\.\/)+`

sarjon replied:
thanks, im not really good at regex. :(

PierreRambaud commented on src/Adapter/Category/Image/CategoryThumbnailRetriever.php:
```diff
@@ -0,0 +1,104 @@
+
+        preg_match('@src="([^"]+)"@', $imageTag, $path);
+
+        return $path[1];
```
make sure $path[1] existe, can happened if src is empty (due to + instead of *)

PierreRambaud commented on src/Adapter/Category/QueryHandler/GetCategoryForEditingHandler.php:
```diff
@@ -0,0 +1,197 @@
+            ImageManager::resize(_PS_TMP_IMG_DIR_ . 'category_' . $categoryId->getValue() . '-thumb.jpg', _PS_TMP_IMG_DIR_ . 'category_' . $categoryId->getValue() . '-thumb.jpg', (int) $image_type['width'], (int) $image_type['height']);
+        }
+
+        $thumbSize = file_exists($thumb) ? filesize($thumb) / 1000 : false;
```
should be 1024 with filesize no?

sarjon replied:
nope, 1024 would be for `KiB` and 1000 is for `kB`.

https://i.imgur.com/faKgB6e.png

i wanted to introduce some kind of size converter as its pretty common task, something like:
`$converter->convert(new Bytes(26589), SIZE::KILOBYTE)`

or `(new Bytes(26589))->toKilobytes()`

jolelievre replied:
this is so fucked up :P I got enraged when I learnt (only 8 months ago..) that this was the official conversion 😔
or how the Disk sellers managed to lobby brain the International System of Unit...

PierreRambaud replied:
@jolelievre so true, that's why I prefer the usage of kiB instead of kB  ^^

PierreRambaud commented on src/Adapter/Category/QueryHandler/GetCategoryForEditingHandler.php:
```diff
@@ -0,0 +1,197 @@
+    {
+        $menuThumbnails = [];
+
+        for ($i = 0; $i < 3; ++$i) {
```
Why 3?

sarjon replied:
its maximum number of menu thumbnails category can have, i cant think of a good place to put this constant? somewhere under `Domain/Category` :thinking: ? @matks wdyt?

PierreRambaud replied:
:+1: for const even if I don't understand why there is a limit :smile:

PierreRambaud commented on src/Adapter/Image/Uploader/CategoryMenuThumbnailUploader.php:
```diff
@@ -0,0 +1,109 @@
+            }
+        }
+
+        \Tools::clearSmartyCache();
```
Is there a Tools adapter instead?

PierreRambaud commented on src/Core/Domain/Category/Command/AbstractRootCategoryCommand.php:
```diff
@@ -0,0 +1,363 @@
+    /**
+     * @var string[]
+     */
+    private $description;
```
same as @matks comments, so many @string[] but not in plurial.

PierreRambaud commented on src/Core/Domain/Category/Command/AbstractRootCategoryCommand.php:
```diff
@@ -0,0 +1,363 @@
+     *
+     * @throws CategoryConstraintException
+     */
+    public function setName(array $name)
```
setNames?

PierreRambaud commented on src/Core/Domain/Category/Command/AbstractRootCategoryCommand.php:
```diff
@@ -0,0 +1,363 @@
+
+        return $this;
+    }
+}
```
I'm really confused with this classe, many array without plurial :/

PierreRambaud commented on src/Core/Domain/Category/Command/AddCategoryCommand.php:
```diff
@@ -0,0 +1,87 @@
+     *
+     * @throws CategoryConstraintException
+     */
+    public function __construct(array $name, array $linkRewrite, $isActive, $parentCategoryId)
```
same as previous comments, array $names or string $name? A category can have many names?

sarjon replied:
same as above, it depends on how you think about it, is it single name in different languages or is it different names?

PierreRambaud replied:
In both situation I think if you have a name in different languages, you have many languages, so, many names :/

PierreRambaud commented on src/Core/Image/Parser/ImageTagSourceParser.php:
```diff
@@ -0,0 +1,70 @@
+            return null;
+        }
+
+        return sprintf('%s%s', $this->shopRootUri, $path[1]);
```
make sure $path[1] exists.

PierreRambaud commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/categories.yml:
```diff
@@ -0,0 +1,31 @@
+admin_category_add:
+    path: /new
+    methods: [GET, POST]
```
GET and POST?

PierreRambaud commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/categories.yml:
```diff
@@ -0,0 +1,31 @@
+
+admin_category_add_root:
+    path: /new-root
+    methods: [GET, POST]
```
GET and POST?

PierreRambaud commented on src/PrestaShopBundle/Resources/config/services/adapter/category.yml:
```diff
@@ -0,0 +1,50 @@
+    prestashop.adapter.category.command_handler.add_category_handler:
+        class: 'PrestaShop\PrestaShop\Adapter\Category\CommandHandler\AddCategoryHandler'
+        parent: 'prestashop.adapter.category.command_handler.abstract_category_handler'
+        public: true
```
default is set to public, any reason to keep it here?

sarjon replied:
yeah, when you use `parent` you must explicitly set it to `public` otherwise is thrown as it cannot be inherited from `_defaults`.

PierreRambaud commented on src/PrestaShopBundle/Resources/config/services/adapter/category.yml:
```diff
@@ -0,0 +1,50 @@
+
+    prestashop.adapter.category.query_handler.get_category_for_editing:
+        class: 'PrestaShop\PrestaShop\Adapter\Category\QueryHandler\GetCategoryForEditingHandler'
+        arguments:
```
if you want to be consistent, remove default public and set public: true here or remove public: everywhere

sarjon replied:
it does not have `parent` so it inherits from `_defaults`

jolelievre replied:
agreed, although there's a slight inconsistency we want to keep the default public value, and set it manually only when required

PierreRambaud commented on src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:
```diff
@@ -551,3 +551,53 @@ services:
+        arguments:
+            - "@=service('prestashop.adapter.form.choice_provider.shop_tree_choice_provider').getChoices()"
+        tags:
+            - { name: form.type }
```
This file becomes and will be bigger and bigger, should we split this one?

matks replied:
Yes ^^ 600 lines is enough indeed

sarjon replied:
:+1: but not in this PR ^^

PierreRambaud commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig:
```diff
@@ -0,0 +1,192 @@
+ *#}
+
+{{ form_start(categoryForm) }}
+<div class="card">
```
You must add twig block everywhere =)

PierreRambaud commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/add.html.twig:
```diff
@@ -0,0 +1,45 @@
+{% block javascripts %}
+  {{ parent() }}
+
+  <script src="{{ asset('themes/new-theme/public/categories.bundle.js') }}"></script>
```
could be a partial too no?

sarjon committed:
use template strings in name-to-link-rewrite-copier.js

sarjon committed:
throw exception when adding / editing fails

sarjon committed:
remove unused image retriever

jolelievre commented on src/Adapter/Category/CommandHandler/AbstractCategoryHandler.php:
```diff
@@ -0,0 +1,142 @@
+        // Inside Category::add() & Category::update() method it checks if shop association is submitted
+        // by retrieving data directly from $_POST["checkBoxShopAsso_category"].
+        $_POST['checkBoxShopAsso_category'] = $command->getAssociatedShopIds();
+
```
I have an idea of the answer, but is it too complicated (worth it, and safe) to change this behaviour in the Category object?

sarjon replied:
personally, im too scared to touch object model. :fearful:

jolelievre replied:
and you should be ^^

sarjon committed:
camel case variable name

jolelievre commented on src/Adapter/Category/CommandHandler/EditCategoryHandler.php:
```diff
@@ -0,0 +1,68 @@
+            $category->id_parent = $command->getParentCategoryId();
+        }
+
+        $this->populateCategoryWithCommandData($category, $command);
```
I thought `populateCategoryWithCommandData` was meant to build root/simple categories, not save them At least that's what is written in the doc from the Abstract class
So either it should not be used here or the doc is incorrect?

sarjon replied:
outdated docs, fixed. ^^

jolelievre commented on src/Adapter/Category/CommandHandler/EditRootCategoryHandler.php:
```diff
@@ -0,0 +1,57 @@
+            );
+        }
+
+        $this->populateCategoryWithCommandData($category, $command);
```
same comment as above

jolelievre commented on src/Adapter/Category/Image/CategoryThumbnailRetriever.php:
```diff
@@ -0,0 +1,104 @@
+     *
+     * @return string
+     */
+    public function retrieve($categoryId)
```
could you add a little comment here or in the class description, about what it is doing
It seems to be converting a path (if I get it correctly), just an example of the initial value and the new one would be good
Or maybe I got it wrong and it justifies a documentation even more ^^

jolelievre replied:
Also, some tests for this class would be good, but maybe you planned to add them in another PR?

matks commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/CategoryType.php:
```diff
@@ -0,0 +1,66 @@
+            ->setDefaults([
+                'id_category' => null,
+            ])
+            ->setAllowedTypes('id_category', ['int', 'null'])
```
I did not know this works 😮 that's cool ! 👍

matks commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig:
```diff
@@ -0,0 +1,192 @@
+          </small>
+        </div>
+      </div>
+      {% if categoryForm.id_parent is defined %}
```
I like to have blank lines between blocks to help me read them however I am aware there are people against it. Wdyt ?

matks commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig:
```diff
@@ -0,0 +1,192 @@
+          {{ form_widget(categoryForm.active) }}
+          <small class="form-text">
+            {{ 'If you want a category to appear in the menu of your shop, go to [1]Modules > Modules & Services > Installed modules.[/1] Then, configure your menu module.'|trans({
+              '[1]': '<a href="' ~ path('admin_module_manage') ~ '" class="_blank">',
```
I did not know this syntax 😮

matks commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig:
```diff
@@ -0,0 +1,192 @@
+      <div class="form-group row">
+        {{ ps.label_with_help(('Category Cover Image'|trans), ('This is the main image for your category, displayed in the category page. The category description will overlap this image and appear in its top-left corner.'|trans({}, 'Admin.Catalog.Help'))) }}
+        <div class="col-sm">
+          {% if coverImage is defined and coverImage is not null %}
```
This can be encapsulated into a twig block

matks commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/edit.html.twig:
```diff
@@ -0,0 +1,50 @@
+{% block content %}
+  <div class="row justify-content-center">
+    <div class="col">
+      {% include '@PrestaShop/Admin/Sell/Catalog/Categories/Blocks/form.html.twig' with {
```
👍

matks replied:
I'm thinking ... shall we be more accurate in the naming of our blocks ? `form.html.twig` could become `category-form.html.twig`
This is useful when you search for a twig file 😅 

I'm not 100% sure about this. Wdyt ?

sarjon replied:
yes, agree. :+1:

matks commented on tests/Unit/Core/Image/Parser/ImageTagSourceParserTest.php:
```diff
@@ -0,0 +1,83 @@
+use PHPUnit\Framework\TestCase;
+use PrestaShop\PrestaShop\Core\Image\Parser\ImageTagSourceParser;
+
+class ImageTagSourceParserTest extends TestCase
```
❤️

matks commented:
@sarjon 1st review done !
Nice job :) you've got some feedbacks but considering the number of lines of code, there's not so many of them

Looks like you had a hard time dealing with the legacy images upload / pathfinding system 😭 sorry about that

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig:
```diff
@@ -0,0 +1,192 @@
+        <div class="col-sm">
+          {{ form_widget(categoryForm.active) }}
+          <small class="form-text">
+            {{ 'If you want a category to appear in the menu of your shop, go to [1]Modules > Modules & Services > Installed modules.[/1] Then, configure your menu module.'|trans({
```
Hum, you should wait for the 1.7.5 to be released, because of the Module Manager... the behavior is supposed to remain the same but the wording will be different. We can check that if you want!

sarjon replied:
this PR will be part of 1.7.6 so we can change wording now, just let me know the correct sentence. :)

LouiseBonnard replied:
I would write: _If you want a category to appear in the menu of your shop, go to [1]Modules > Module Manager[/1]. Then, configure your menu module_. @LouToma, can you confirm this path?

LouToma replied:
We can even write "If you want a category to appear in the menu of your shop, go to [1]Modules > Module Manager[/1] and scroll down to the "Design & Navigation" category. Then, configure you Menu module.

LouiseBonnard replied:
Perfect, it is the right path! Let's keep the following wording (_If you want a category to appear in the menu of your shop, go to [1]Modules > Module Manager[/1]. Then, configure your menu module._), thanks.

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig:
```diff
@@ -0,0 +1,192 @@
+              <figure class="figure">
+                <img src="{{ coverImage.path }}" class="figure-img img-fluid img-thumbnail">
+                <figcaption class="figure-caption">
+                  <p>{{ 'File size'|trans }} {{ coverImage.size }}</p>
```
Should be localized in _Admin.Advparameters.Feature_.

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig:
```diff
@@ -0,0 +1,192 @@
+            <div>
+              <figure class="figure">
+                <img src="{{ thumbnailImage.path }}" class="figure-img img-fluid img-thumbnail">
+                <figcaption class="figure-caption">{{ 'File size'|trans }} {{ thumbnailImage.size }}</figcaption>
```
Same here, thanks!

sarjon committed:
update menu thumbnails uploader

sarjon committed:
remove duplicate hook call

jolelievre commented on src/Adapter/Image/Uploader/CategoryCoverImageUploader.php:
```diff
@@ -0,0 +1,136 @@
+    public function upload($id, UploadedFile $uploadedImage)
+    {
+        $this->deleteOldImage($id);
+        $this->checkImageIsAllowedForUpload($uploadedImage);
```
maybe you should check the image before deleting the old one? the safest would even be to delete the old one ONLY when the three other steps suceeded but this might not be possible

sarjon replied:
even though i agree with you, but lets keep behavior as it was in legacy. we can change it in another PR if it must be. :)

jolelievre replied:
well when there is a risk to break something or generating side effects, yeah I would agree But here clearly it was not thought through in the legacy code, and it won't break a desired behavior it will just fix a potential bug
Refacto/migation is not only about doing the same shit in a classier way :P If we can fix small bugs like this one in the process we should

sarjon replied:
> and it won't break a desired behavior it will just fix a potential bug

i've tested it and apparently we cant delele old images after uploading new one. The reason is that both new and old images use same naming, e.g. `{categoryId}-thumb.jpg`. so deleting old ones would delete actually new ones. :P

matks replied:
😱

jolelievre replied:
damn!!! well too bad ^^

jolelievre commented on src/Adapter/Image/Uploader/CategoryCoverImageUploader.php:
```diff
@@ -0,0 +1,136 @@
+     * @throws ImageUploadException
+     * @throws MemoryLimitException
+     */
+    private function uploadImage(UploadedFile $image, $id)
```
better keep the same arguments order as the `upload` method for consistency

jolelievre commented on src/Adapter/Image/Uploader/CategoryMenuThumbnailUploader.php:
```diff
@@ -0,0 +1,109 @@
+     *
+     * @throws ImageUploadException
+     */
+    public function upload($categoryId, UploadedFile $uploadedImage)
```
you don't respect the `ImageUploaderInterface` argument name `$id` either change the interface with a more generic name like `$entityId` or this class

sarjon replied:
i dont know if that can be called _not respecting_ an interface, but i've update name to `$entityId`. :)

jolelievre replied:
I agree with you on the `respect`, as long as type is correct PHP might accept it, it is more about consistency ;)
Thanks anyway

jolelievre commented on src/Core/Domain/Category/Command/AbstractRootCategoryCommand.php:
```diff
@@ -0,0 +1,363 @@
+/**
+ * Class AbstractAddCategoryCommand defines common command for both simple and root category creation.
+ */
+abstract class AbstractRootCategoryCommand
```
Inconsistency between the doc and the class name
Also I don't think `AbstractRootCategoryCommand` because you use it for `AddCategoryCommand, AddRootCategoryCommand, EditCategoryCommand, EditRootCategoryCommand` so pretty much all your commands, maybe you can simply call it `AbstractCategoryCommand`
Finally, as I mentioned earlier, you should update `populateCategoryWithCommandData` doc because you claim this method is used for creation only whereas it is used everywhere, which is not a problem just the doc comment is confusing
And maybe you should use an interface in `populateCategoryWithCommandData, uploadImages` but on this one I am totally not sure since you might want to limit these helpers to the `AbstractRootCategoryCommand` scope and not open it to external commands

sarjon committed:
update selector for text-with-length-counter js

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -0,0 +1,418 @@
+     * @param array $data
+     */
+    protected function populateCommandWithFormData(AbstractRootCategoryCommand $command, array $data)
+    {
```
missing `setName`, `setUrlRewrite`, `setIsActive`, and maybe `setParentCategoryId` ?

jolelievre commented on src/Core/Domain/Category/Command/EditCategoryCommand.php:
```diff
@@ -0,0 +1,95 @@
+     *
+     * @throws CategoryConstraintException
+     */
+    public function setParentCategoryId($parentCategoryId)
```
maybe you could factorize this parentCategoryId management in an `AbstractChildCategoryCommand` which would be common between `EditCategoryCommand` and `AddCategoryCommand`

jolelievre commented on src/Core/Domain/Category/Command/AbstractRootCategoryCommand.php:
```diff
@@ -0,0 +1,363 @@
+     * @var string[]
+     */
+    private $name;
+
```
I would use `$localizedNames` to clarify

jolelievre commented on src/Core/Domain/Category/Command/AbstractRootCategoryCommand.php:
```diff
@@ -0,0 +1,363 @@
+     * @throws CategoryConstraintException
+     */
+    public function setName(array $name)
+    {
```
and of course `getLocalizedNames` / `setLocalizedNames` even though it is not consistent with the `Category` object I think it's the legacy Object which is poorly named, we shouldn't keep bad naming because of legacy code

sarjon committed:
update template with new blocks

jolelievre commented on src/Core/Domain/Category/Command/AbstractRootCategoryCommand.php:
```diff
@@ -0,0 +1,363 @@
+     * @var string[]
+     */
+    private $description;
+
```
same here `$localizedLinkRewrites` and `$localizedDescriptions`

jolelievre commented on src/Core/Domain/Category/Command/AbstractRootCategoryCommand.php:
```diff
@@ -0,0 +1,363 @@
+     * @var string[]
+     */
+    private $metaDescription;
+
```
`$localizedMetaTitles` `$localizedMetaDescriptions`

jolelievre commented on src/Core/Domain/Category/Command/AbstractRootCategoryCommand.php:
```diff
@@ -0,0 +1,363 @@
+    /**
+     * @var string[]
+     */
+    private $metaKeywords;
```
`$localizedMetaKeywords`

jolelievre commented on src/Core/Domain/Category/EditableCategory.php:
```diff
@@ -0,0 +1,259 @@
+ */
+class EditableCategory
+{
+    /**
```
same modification here as in the commands all localized fields need to be renamed `$name` -> `$localizedNames`
this stands for `name, description, metaTile, metaDescription, metaKeywords, linkRewrite`

jolelievre commented on src/Core/Domain/Category/Exception/CategoryConstraintException.php:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Class CategoryConstraintException.
+ */
+class CategoryConstraintException extends CategoryException
```
very nice! as we talked about could you please create `CoreException` (forget about the PrestaShop prefix) in `src/Core/Exception` and rename the existing `PrestaShopCoreException` into a `NotificationException` or maybe `ErrorMessageException` (less linked to the notification feature)

sarjon committed:
extract some blocks into separate templates

jolelievre commented on src/Core/Domain/Category/Exception/CategoryException.php:
```diff
@@ -24,12 +24,13 @@
+/**
+ * Class CategoryException is base exception for Category bounded context.
+ */
+class CategoryException extends DomainException
```
I don't understand what this `DomainException` stands for.. probably a dum question but what is a domain?

jolelievre replied:
nonetheless `DomainException` should extend `CoreException`

jolelievre commented on src/Core/Domain/Group/ValueObject/GroupId.php:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Class GroupId.
+ */
+class GroupId
```
shouldn't all these entity id classes have a common interface like the `EntityIdentifierInterface` we talked about for entity form architecture, but maybe it's too soon

sarjon committed:
wording update

sarjon commented:
huh, i've addressed most of your comments @matks @PierreRambaud :sweat:

jolelievre commented on src/Core/Image/Uploader/Exception/ImageOptimizationException.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Class ImageOptimizationException is thrown when resizing, cutting or optimizing image fails.
+ */
+class ImageOptimizationException extends \Exception
```
should extend at least `CoreException` or maybe `ImageException` or even `ImageUploadException` in this case (both, fo course, extending `CoreException`)

jolelievre commented on src/Core/Image/Uploader/Exception/ImageUploadException.php:
```diff
@@ -0,0 +1,31 @@
+
+namespace PrestaShop\PrestaShop\Core\Image\Uploader\Exception;
+
+class ImageUploadException extends \Exception
```
same here extends `ImageUploadException` (sorry for the repeat, just as a memo)

jolelievre commented on src/Core/Image/Uploader/Exception/MemoryLimitException.php:
```diff
@@ -0,0 +1,31 @@
+
+namespace PrestaShop\PrestaShop\Core\Image\Uploader\Exception;
+
+class MemoryLimitException extends \Exception
```
same here extends `ImageUploadException` (sorry for the repeat, just as a memo)

jolelievre commented on src/Core/Image/Uploader/Exception/UploadedImageConstraintException.php:
```diff
@@ -0,0 +1,34 @@
+
+namespace PrestaShop\PrestaShop\Core\Image\Uploader\Exception;
+
+class UploadedImageConstraintException extends \Exception
```
same here extends `ImageUploadException` (sorry for the repeat, just as a memo)

jolelievre commented on src/Core/Image/Uploader/Exception/UploadedImageConstraintException.php:
```diff
@@ -0,0 +1,34 @@
+{
+    const EXCEEDED_SIZE = 1;
+    const UNRECOGNIZED_FORMAT = 2;
+    const UNKNOWN_ERROR = 4;
```
you should use explicit string values (like `const EXCEEDED_SIZE = 'exceeded_size';`) rather than integers, integer don't hold any meaning when you see them in a stack trace, a log or any other printed output
integer values with power of 2 values like you used here only have interest when you need to cumulate errors via binary addition (e.g.  EXCEEDED_SIZE | UNRECOGNIZED_FORMAT) but here it seems to me that each type is used in different situations

jolelievre replied:
ok, I just understood that you use these values for the \Exception::$code value which is an int so actually you can't use string here, my bad
I'm just not a fan of using the same exception for different purposes, because you need to perform a test on the code in your catch block, but it's not something I can force on other
Good job!

sarjon replied:
maybe different exception for each error type?

jolelievre replied:
that would be the best sure, maybe a bit overkill to create 4 exceptions (one as a base and three for each sub cases)? hard to arbitrate here..

matks replied:
It is interesting to have 4 exceptions class if you are going to write code based on that, either
```
try {
   ...
} catch (Exception1 $e) {
    ...
}
// let Exception 2 go through
```
or
```
} catch (Exception $e) {
    $class = get_class($e);
    if ($class === Exception1::class) {
        ...
    }
}
```
If we do not use them, let's stick to the codes ;)

jolelievre replied:
as @eternoendless would say: "that's not incorrect" ^^

mickaelandrieu replied:
not sure it's possible but how about:

```php
try
{
    // Some code...
}
catch(Exception1 | Exception2 | Exception3 $e )
{
    
}
```

jolelievre commented on src/Core/Domain/Category/Exception/CategoryConstraintException.php:
```diff
@@ -0,0 +1,63 @@
+     * Code is used when category does not have name.
+     */
+    const EMPTY_NAME = 1;
+
```
sorry I noticed this one later, you should use string rather than integers
you'll find more details in the `UploadedImageConstraintException` comment

jolelievre commented on src/Adapter/Image/Uploader/AbstractImageUploader.php:
```diff
@@ -0,0 +1,77 @@
+        $maxFileSize = Tools::getMaxUploadSize();
+
+        if ($maxFileSize > 0 && $image->getSize() > $maxFileSize) {
+            throw new UploadedImageConstraintException(
```
maybe here you should use the `ErrorMessageException` discussed in another comment, this way you could store your `key`, `parameters` and especially `domain` in it
(the domain is lacking with this current Exception)
I think maybe the use of ErrorMessageException as a base should be generalized in any Exception that holds parameter, because if you inject the parameters directly in the Exception message you can't translate them any more
This is not a problem for messages without parameters because the exceptions can be caught elsewhere and then the controller can translate it with the proper domain

Which is why I think in `ErrorMessageException` both `parameters` and `domain` should be optional, this way you can can throw an Exception with the key and parameters without caring about the domain which you might not know at the moment

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -0,0 +1,418 @@
+                );
+                $this->populateCommandWithFormData($command, $data);
+
+                $this->getCommandBus()->handle($command);
```
I'm wondering here about how we allow developer to override / add feature to the core code
Should we have a hook before handling the command? so that you can change or add information?
Or can it be managed in an CommandHandler?

I mean what if I want to add an extra field to my Categories and I need it on creation, then the way it is handled here I can't add it even though I managed to modify my form (see later in EntityForm which might need a FormDefinition like the Grid)

Just a question here, no modifications needed for this PR as it should be dealt with later.

sarjon replied:
i dont really have an answer here but my guess is, we either:
1. add middleware for command bus that allows that (maybe dispatches hooks)
2. make form handler (which does not yet exist) dispatch hook before/after executing command with all form data.

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -0,0 +1,418 @@
+            'shop_association' => [
+                $this->getContextShopId(),
+            ],
+        ];
```
maybe you can factorize this part in a private method since it is exactly the same as in addAction, a `buildEmptyCategoryData($defaultGroups)` method
although the gain is not that important..

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -0,0 +1,418 @@
+
+        /** @var DefaultGroups $defaultGroups */
+        $defaultGroups = $this->getQueryBus()->handle(new GetDefaultGroups());
+
```
you already fetch $defaultGroups above

jolelievre commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/CategoryType.php:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Class CategoryType.
+ */
+class CategoryType extends AbstractCategoryType
```
maybe call this `CategoryChildType` as opposition to `CategoryRootType`

sarjon replied:
i think we dont have `Category child` in Back Office, it's only `Category` or `Root Category`, naming things differently in code may be confusing. :/

jolelievre replied:
fair point!

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -0,0 +1,418 @@
+            'group_association' => $editableCategory->getGroupAssociationIds(),
+            'shop_association' => $editableCategory->getShopAssociationIds(),
+        ];
+
```
maybe a `toArray()` method in `EditableCategory`?
although I am not sure it is its responsibility to format data form the Form

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -0,0 +1,418 @@
+
+                if (null !== $data['id_parent']) {
+                    $command->setParentCategoryId($data['id_parent']);
+                }
```
are there some cases where `id_parent` is null? I though only root category had no parent, and they have their own controller
or is it when you want to transform a child category into a root category?

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -0,0 +1,418 @@
+                [
+                    '%value%' => $editableCategory->getName()[$this->getContextLangId()],
+                ]
+            ),
```
I am not sure, should this be managed in the controller or in the template?

jolelievre commented on src/PrestaShopBundle/Form/Admin/Type/TranslatableType.php:
```diff
@@ -0,0 +1,104 @@
+    {
+        $view->vars['locales'] = $options['locales'];
+        $view->vars['default_locale'] = reset($options['locales']);
+        $view->vars['hide_locales'] = 1 >= count($options['locales']);
```
could you write it `count($options['locales']) <= 1` which seems easier to read to a human

jolelievre commented on src/PrestaShopBundle/Form/Admin/Type/TranslateTextareaType.php:
```diff
@@ -0,0 +1,89 @@
+use Symfony\Component\Form\FormView;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class TranslateTextareaType extends TranslatorAwareType
```
I think you could extend `TranslatableType` for this one?

jolelievre commented on src/PrestaShopBundle/Resources/config/services/adapter/category.yml:
```diff
@@ -0,0 +1,50 @@
+        parent: 'prestashop.adapter.category.command_handler.abstract_category_handler'
+        public: true
+        tags:
+            - { name: 'tactician.handler', command: 'PrestaShop\PrestaShop\Core\Domain\Category\Command\AddCategoryCommand' }
```
be consistent in your writing, either write on one line or two lines But use the same syntax in the whole file

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig:
```diff
@@ -0,0 +1,192 @@
+ * International Registered Trademark & Property of PrestaShop SA
+ *#}
+
+{{ form_start(categoryForm) }}
```
I think you should declare a block here for override purposes, but sadly we have no guidelines for this for now
So maybe don't bother for now

sarjon committed:
update wording

jolelievre reviewed (dismissed):
woooow, that was a long one!! ^^
to be honest I skipped though the last twig templates.. :P
Really nice job overwhole!! Congrats @sarjon this is a really good base for CQRS I think, I learnt a lot through it! Thank you

sarjon committed:
remove TranslateTextareaType

sarjon committed:
update AbstractCategoryHandler docs

sarjon committed:
update argument order

sarjon committed:
update ImageUploaderInterface docs and naming

sarjon committed:
update commands doc blocks

sarjon commented:
@matks can we address remaining comments in separate PR with other changes that needs to be done for categories form?

PierreRambaud commented on src/Adapter/Category/QueryHandler/GetCategoryForEditingHandler.php:
```diff
@@ -0,0 +1,197 @@
+        $imageSize = file_exists($image) ? filesize($image) / 1000 : '';
+
+        if (empty($imageTag) || empty($imageSize)) {
+            return null;
```
The doc says it must return an array. Throw exception, return an empty array or change doc block please :)

PierreRambaud commented on src/Adapter/Category/QueryHandler/GetCategoryForEditingHandler.php:
```diff
@@ -0,0 +1,197 @@
+        }
+
+        return [
+            'size' => sprintf('%skb', $thumbSize),
```
`kB`

PierreRambaud commented on src/Adapter/Image/Uploader/AbstractImageUploader.php:
```diff
@@ -0,0 +1,77 @@
+
+        if (!ImageManager::isRealImage($image->getPathname(), $image->getClientMimeType())
+            || !ImageManager::isCorrectImageFileExt($image->getClientOriginalName())
+            || preg_match('/\%00/', $image->getClientOriginalName())
```
:+1: Maybe add a comment to says it is to prevent the null byte injection (even if this code comes from legacy)

sarjon replied:
i didnt know that to be honest.. ^^

PierreRambaud replied:
Little explanation: http://www.madirish.net/401

sarjon replied:
thanks. :)

sarjon committed:
minor updates

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/category_form.html.twig:
```diff
@@ -0,0 +1,168 @@
+      </div>
+
+      <div class="form-group row">
+        {{ ps.label_with_help(('Displayed'|trans({}, 'Admin.Global')), ('Click on "Displayed" to index the category on your shop.'|trans({}, 'Admin.Catalog.Help'))) }}
```
I have got issues on Crowdin with this wording, people ask for the use of the present tense and rephrasing with suggestions like _Click on "Display" if you want the category to appear in your shop front._ or _Click on "Display" to index the category on your shop._, perhaps it is time to take the opportunity of the migration and think about a. better formulation? @colinegin, do you think that we could even think about deleting the mention of the verb and mentioned it as a mere "button"?

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/category_form.html.twig:
```diff
@@ -0,0 +1,168 @@
+      </div>
+
+      <div class="form-group row">
+        {{ ps.label_with_help(('Meta keywords'|trans({}, 'Admin.Global')), ('To add "tags," click in the field, write something, and then press "Enter."'|trans({}, 'Admin.Catalog.Help')) ~ ('Forbidden characters:'|trans({}, 'Admin.Notifications.Info')) ~ ' <>;=#{}') }}
```
Second string contains two typos, it should be written as follows: _To add tags, click in the field, write something, and then press the "Enter" key._. Otherwise, quotation marks are misplaced. And this string is localized in _Admin.Shopparameters.Help_.

sarjon replied:
actually string `To add "tags," click in the field, write something, and then press "Enter."` is also available under `Admin.Catalog.Help` domain. But i can update according you suggestion. :+1:

LouiseBonnard replied:
I know, you're right, it already exists in the catalog, but when you look closer you can see that the quotation marks are not in the right place, this is why there are twice the same string. And then I prefer to keep the right one. ;-)

sarjon committed:
fix categories for all shop context

sarjon committed:
minor update

sarjon committed:
move layout titles to templates

sarjon committed:
use bool instead of int for root category

sarjon committed:
cs fixer

sarjon committed:
update wording

sarjon committed:
update naming

sarjon committed:
handle image upload during category edit

matks commented:
@sarjon Yes ! Let's merge this PR and create another one to handle the last feedbacks.

I see Travis build is red. But the errors do not look related to this PR so I restart it.

sarjon commented:
@matks remember that this does not implement whole functionality for add/edit form, some are still missing. So QA wont be able to validate everything. :/ maybe QA can validate when its fully migrated in order to avoid confusion?

matks commented:
@sarjon Yes indeed. Let's merge it and QA will validate the 2nd PR which will complete this work.

Travis build is red again 😢 I'm gonna check this locally.

matks commented:
I fail to understand why the Travis tests fail 😭

sarjon commented:
@matks ill try to debug them locally. :)

sarjon committed:
fix rebase issues

mickaelandrieu commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/category_form.html.twig:
```diff
@@ -0,0 +1,168 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ *#}
+
```
you didn't have imported the macro, this is required for Twig 2

sarjon replied:
hope this will work. :crossed_fingers:

sarjon committed:
include macros into template

sarjon commented:
@mickaelandrieu i've imported macros, but tests are still failing. :cry:

matks commented:
🔥

PierreRambaud commented:
@sarjon @matks currently trying to reproduce.

PierreRambaud committed:
Test can be broken on certain platform

The scandir use SCANDIR_SORT_NONE, to prevent random order only check
for array values and array length.

PierreRambaud committed:
Try to change StockManagement Tests

PierreRambaud commented:
@sarjon this one needs to be rebased or if you don't want to waste 2hours you can directly pull develop.  
Travis can't run tests because not mergeable:
```
GitHub payload is missing a merge commit (mergeable_state: "unknown", merged: false)
```

PierreRambaud commented:
I'm done with these tests :'(
@sarjon You can revert my commit `3e5840c` please? Look like it's not working :/ 
`990753d` needs to be keept

sarjon committed:
Revert "Try to change StockManagement Tests"

This reverts commit 3e5840cb69ff69a9e754d63e6dc2e18399822eed.

sarjon committed:
minor code formatting fixes

sarjon committed:
build assets

mickaelandrieu commented on src/Core/Domain/Group/Query/GetDefaultGroups.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Class GetDefaultGroups returns default groups data.
+ */
+class GetDefaultGroups
```
empty object? if a Query is expected, why we dont have an interface for it? :)

mickaelandrieu commented on src/Core/Image/Parser/ImageTagSourceParser.php:
```diff
@@ -1,6 +1,14 @@
 <?php
 /**
+<<<<<<< HEAD
```
conflicts here

matks commented:
@mickaelandrieu As this PR grew too big and too old, we're going to merge it then do a 2nd one which will handle your feedbacks, other code review feedbacks and QA validation 😄
